ambry-router/src/test/java/com.github.ambry.router/DefaultGCMCryptoServiceTest.java

```diff
@@ -0,0 +1,151 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.utils.TestUtils;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.util.Arrays;
+import java.util.Properties;
+import javax.crypto.spec.SecretKeySpec;
+import org.bouncycastle.util.encoders.Hex;
+import org.junit.Assert;
+import org.junit.Test;
+
+import static com.github.ambry.router.CryptoTestUtils.*;
+
+
+/**
+ * Tests {@link DefaultGCMCryptoService} and {@link DefaultCryptoServiceFactory}
+ */
+public class DefaultGCMCryptoServiceTest {
+
+  private static final int MAX_DATA_SIZE = 10000;
+  private static final int DEFAULT_KEY_SIZE_IN_CHARS = 64;
+
+  /**
+   * Tests basic encryption and decryption for different sizes of keys and random data in bytes
+   */
+  @Test
+  public void testEncryptDecryptBytes() throws Exception {
+    for (int j = 0; j < 3; j++) {
+      String key = getRandomKey(DEFAULT_KEY_SIZE_IN_CHARS);
+      Properties props = getKMSProperties(key, DEFAULT_KEY_SIZE_IN_CHARS);
+      VerifiableProperties verifiableProperties = new VerifiableProperties((props));
+      SecretKeySpec secretKeySpec = new SecretKeySpec(Hex.decode(key), "AES");
+      CryptoService<SecretKeySpec> cryptoService =
+          new DefaultCryptoServiceFactory(verifiableProperties).getCryptoService();
+      for (int i = 0; i < 5; i++) {
+        int size = TestUtils.RANDOM.nextInt(MAX_DATA_SIZE);
+        byte[] randomData = new byte[size];
+        TestUtils.RANDOM.nextBytes(randomData);
+        ByteBuffer toEncrypt = ByteBuffer.wrap(randomData);
+        ByteBuffer encryptedBytes = cryptoService.encrypt(toEncrypt, secretKeySpec);
+        Assert.assertFalse("Encrypted bytes and plain bytes should not match",
+            Arrays.equals(randomData, encryptedBytes.array()));
+        ByteBuffer decryptedBytes = cryptoService.decrypt(encryptedBytes, secretKeySpec);
+        Assert.assertArrayEquals("Decrypted bytes and plain bytes should match", randomData, decryptedBytes.array());
+      }
+    }
+  }
+
+  /**
+   * Tests basic encryption and decryption for different sizes of keys and random data
+   */
+  @Test
+  public void testEncryptDecryptKeys() throws Exception {
+    for (int j = 0; j < 5; j++) {
+      String keyToEncrypt = getRandomKey(DEFAULT_KEY_SIZE_IN_CHARS);
+      Properties props = getKMSProperties(keyToEncrypt, DEFAULT_KEY_SIZE_IN_CHARS);
+      VerifiableProperties verifiableProperties = new VerifiableProperties((props));
+      SecretKeySpec secretKeyToEncrypt = new SecretKeySpec(Hex.decode(keyToEncrypt), "AES");
+      String keyToBeEncrypted = getRandomKey(DEFAULT_KEY_SIZE_IN_CHARS);
+      SecretKeySpec secretKeyToBeEncrypted = new SecretKeySpec(Hex.decode(keyToBeEncrypted), "AES");
+      CryptoService<SecretKeySpec> cryptoService =
```

Comment:
minor: looks like an instance of this can be created in the constructor (which is how most users will use it - create a single instance and use it repeatedly)

Reply:
I want to test using different keys and most of the tests are run with N times (with diff keys each time)